### PR TITLE
History singletons

### PR DESCRIPTION
**Pull request by Schniz**
Instead of just updating the docs, this is an implementation for what @ryanflorence said on #1339 and @mjackson on #1426.

this PR lets you use a history singleton with `BrowserHistory` and `HashHistory`:
```js
var { history } = require('react-router/lib/HashHistory');
React.render(<Router history={ history } ... />, ...);
```

### CODE REVIEW

**Comment by Schniz** [comment]
@mjackson, I decided on calling it `history` instead of `getSingleton`.
Its less about implementation detail (that can be changed later), and more about what you get from this function

File: modules/BrowserHistory.js [hunk 1]
```diff
@@ -106,6 +108,10 @@ class BrowserHistory extends DOMHistory {
     }
   }
 
+  static get history() {
```
Comment (mjackson) [inline modules/BrowserHistory.js]: People who need to support IE8 (poor folks) are going to gripe about this. Let's just `export history new BrowserHistory` in this module instead. Then people can still `import { history }` and we'll be fine.

File: modules/HashHistory.js [hunk 2]
```diff
@@ -168,6 +169,10 @@ class HashHistory extends DOMHistory {
     return '#' + path;
   }
 
+  static get history() {
```
Comment (mjackson) [inline modules/HashHistory.js]: Same.

**Comment by mjackson** [comment]
Thanks for your work on this @Schniz !

**Comment by gaearon** [comment]
💃

**Comment by Schniz** [comment]
damn you IE 8.

**Comment by Schniz** [comment]
@mjackson fixed this :+1:

File: modules/BrowserHistory.js (hunk 1 above)
Comment (Schniz) [inline modules/BrowserHistory.js]: :+1:

**Comment by Schniz** [comment]
the tests don't pass for some reason, and don't seem to be mine. should I rebase for the current master?

**Comment by mjackson** [comment]
@Schniz The only failing test is the one that tests scrolling, which @taurose fixed in c61a0f5f1392afa3d46a4ac05908fa8a4ef16b9d. We're currently working on getting that branch merged, so this one should be good to go for now. When that branch merges we'll be green again.

Thanks!

**Pull request merged**